### ISSUE TITLE
Updating function overrides dependsOn and causes push to fail

### ISSUE THREAD

**Issue by idanlo** [aws-amplify/amplify-cli#6214]
**Describe the bug**
After doing `amplify function update` and updating some dependencies it rewrites the dependsOn in `backend-config` and `amplify-meta` JSON files to be incorrect, then when pushing you get this error - 
```
UPDATE_FAILED               AmplifyResourcesPolicy                                                   AWS::IAM::Policy           Fri Dec 18 2020 13:00:50 GMT+0200 (Israel Standard Time) The policy failed legacy parsing (Service: AmazonIdentityManagement; Status Code: 400; Error Code: MalformedPolicyDocument; Request ID: XXXXXX; Proxy: null)
```

**Amplify CLI Version**
4.40.0

**Expected behavior**
I am able to update my function and keep the existing dependencies.

**Screenshots**
This is a screenshot of the git diff after updating the function dependencies - 
![image](https://user-images.githubusercontent.com/28350511/102608272-4f53e300-4132-11eb-8443-9cd0e2181bab.png)

**Comment by yuth** [comment]
@idanlo Could you include more details about the changes that you did, like what kinda of resources the function had access to prior to update and what new resources were granted access to from the function

**Comment by idanlo** [comment]
@yuth this actually just happened again.
I was using cli version 4.37.1 because the new versions are not working with multiple environments, so with version 4.37.1 I tried adding access to an S3 bucket for a lambda function, this function already has multiple resources connected to it such as auth, api, analytics. After adding access to an S3 bucket nothing changed in the backend-config file and the amplify-meta file - nothing at all, I pushed the function successfully and later found out that the environment variable in the lambda for the bucket name is just a placeholder, so that push shouldn't have even worked. After upgrading back to 4.40.1 I tried to add the permission for the function again, and now it just deleted all the `dependsOn` in backend-config and amplify-meta and left the `auth` category there, very weird behavior, if I were to push my function, all my resources would now not work. This is frustrating that I have to manually change it every time I update permissions for any of my functions

**Comment by idanlo** [comment]
Also after downgrading back to 4.37.1 and updating permissions again, doesn't affect the backend-config and amplify-meta, they just stay with just the `auth` dependency

**Comment by edwardfoyle** [comment]
Hi @idanlo I was unable to reproduce this with the latest version of the CLI. On this question:
```
? Which setting do you want to update? Resource access permissions
? Select the category (Press <space> to select, <a> to toggle all, <i> to in
vert selection)
❯◉ auth
 ◉ api
 ◉ storage
 ◯ function
```
even if you only want to update "auth" you still need to leave all the options selected as the CLI will interpret deselecting categories as removing all permissions for those categories. The wording of this prompt makes this confusing though.

Can you confirm if you are only selecting auth and that's what's causing the other permissions to be removed?

**Comment by Stvad** [comment]
I'm experiencing the same/similar issue. The function has access to dynamo table after initial setup, but when I grant it access to auth - that fails in stack update, because it removes access to dynamo table.  

That prompt is confusing (as it implies selecting only things you want to update). But! For me it also by default has nothing selected, so unless you re-select old permissions manually - they get removed

**Comment by renebrandel** [comment]
@Stvad - we've updated the CLI flow to accommodate that. Can you update to the latest CLI version and reproduce this issue again?

**Comment by pedramp20** [comment]
I am experiencing the same issue with CLI 4.44.1 and 4.44.2. Basically, when function resource access is updated and for my case api is added to the function via **amplify update function**, cli removes depends on of the of the existing access from _backend-config.json_ and causes update to fail and throws the following error:

```
The policy failed legacy parsing (Service: AmazonIdentityManagement; Status Code: 400; Error Code: MalformedPolicyDocument; Request ID: xxxx-xxxxx-xxxxx; Proxy: null)
```

**Comment by akshbhu** [comment]
Hi @idanlo 

This is happening due to this :
https://github.com/aws-amplify/amplify-cli/blob/27030eaae7ad83b32af18aa448743a994eb23d9d/packages/amplify-category-function/src/provider-utils/awscloudformation/service-walkthroughs/execPermissionsWalkthrough.ts#L229

Since we are adding dependsOn only on basis for resourceName and in your case analytics and api have same resourceNames which skips the dependsOn for api.

I have included this change in PR #6639

For the workaround , you need to add the dependsOn section in your amplify-meta.json in functions of categories which has same respurceNames
```
      "dependsOn": [
        {
          "category": "api",
          "resourceName": "translatedev",
          "attributes": [
            "GraphQLAPIIdOutput",
            "GraphQLAPIEndpointOutput"
          ]
        }
```
